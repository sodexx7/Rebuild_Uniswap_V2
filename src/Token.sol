// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.21;

// import {ERC20} from "solmate/tokens/ERC20.sol";

import {ERC20} from "openzeppelin-contracts/contracts/token/ERC20/ERC20.sol";

contract Token is ERC20 {
    constructor() ERC20("TestERC20", "TTERC") {}
}
