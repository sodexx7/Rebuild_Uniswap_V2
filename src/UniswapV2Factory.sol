// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.21;

import "./interfaces/IUniswapV2Factory.sol";
import "./UniswapV2Pair.sol";

contract UniswapV2Factory is IUniswapV2Factory {
    address public feeTo;
    address public feeToSetter;
    mapping(address => mapping(address => address)) public getPair;
    address[] public allPairs;

    constructor(address _feeToSetter) public {
        feeToSetter = _feeToSetter;
    }

    function allPairsLength() external view returns (uint256) {
        return allPairs.length;
    }

    function createPair(address tokenA, address tokenB) external returns (address pair) {
        require(tokenA != tokenB, "UniswapV2: IDENTICAL_ADDRESSES");
        (address token0, address token1) = tokenA < tokenB ? (tokenA, tokenB) : (tokenB, tokenA);
        // 1: token0 not equal 0, and because token 1> token0, so token1>token0 NOT euqal 0
        require(token0 != address(0), "UniswapV2: ZERO_ADDRESS");
        require(getPair[token0][token1] == address(0), "UniswapV2: PAIR_EXISTS"); // single check is sufficient
        // bytes memory bytecode = type(UniswapV2Pair).creationCode;
        bytes memory bytecode = abi.encodePacked(type(UniswapV2Pair).creationCode, abi.encode("Uniswap V2", "UNI-V2"));
        bytes32 salt = keccak256(abi.encodePacked(token0, token1));
        assembly {
            pair := create2(0, add(bytecode, 32), mload(bytecode), salt)
        }
        IUniswapV2Pair(pair).initialize(token0, token1);
        getPair[token0][token1] = pair;
        getPair[token1][token0] = pair; // populate mapping in the reverse direction
        allPairs.push(pair);
        emit PairCreated(token0, token1, pair, allPairs.length);
    }

    function setFeeTo(address _feeTo) external {
        require(msg.sender == feeToSetter, "UniswapV2: FORBIDDEN");
        feeTo = _feeTo;
    }

    function setFeeToSetter(address _feeToSetter) external {
        require(msg.sender == feeToSetter, "UniswapV2: FORBIDDEN");
        feeToSetter = _feeToSetter;
    }

    // create pair
    // 0 check
    // 1. check the two address can't be equal
    // 2. check the two address can't be zero address
    // 3. check the two address can't have be used for this pair. have been to make a pair, the trick: when judge the getPair[token0][token1] == address(0) also judge the
    //  getPair[token1][token0] == address(0), because all set while creating the pair
    // 4.

    //  1 make pair
    // 1. create2 code
    //   by create2 code
    //      the difference between create2 and create ???
    //
    // question
    //  abi.encodePacked(token0, token1)   address token0, address token1, the same as abi.encode(token0, token1)  address 160bits. 320bits?
    //  abi.encodePacked, truncate the tail.

    // 1. create2, pehaps have some same address? 2. sometime the same salt?
    // 2.

    // 2. quesiton 2, in my understanding, the below contract can't be executed, because the contract was contracted and in one transaction ,and the transaction not end.
    /**
     *   assembly {
     *         pair := create2(0, add(bytecode, 32), mload(bytecode), salt)
     *     }
     *     IUniswapV2Pair(pair).initialize(token0, token1);
     */

    //

    // https://www.evm.codes/#F5 doing??

    // other consideration confirm
    // the original desgin consideration
    //  1: why need cache token0 and token1???
    //  2:

    // summary
    // createPari
    //  uniswapFactory main funciton, create pair and store all the pair, by array  address[] public allPairs;
    // 1. create pair address, which specify the two token address. only the two token address.
    //  how to get the pair address  getPair[token0][token1]
    //
}
