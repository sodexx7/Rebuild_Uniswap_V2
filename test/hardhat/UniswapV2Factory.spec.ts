import { Wallet,BigNumber, utils } from 'ethers'
import { ethers, waffle } from 'hardhat'
import {UniswapV2Factory} from "../../typechain-types/src/UniswapV2Factory"

import { expect } from './shared/expect'
import snapshotGasCost from './shared/snapshotGasCost'


import {getCreate2Address } from './shared/utilities'
const { constants } = ethers
import foundry_UniswapV2Pair from './shared/foundry_UniswapV2Pair.json'

const TEST_ADDRESSES: [string, string] = [
  '0x1000000000000000000000000000000000000000',
  '0x2000000000000000000000000000000000000000'
]

const createFixtureLoader = waffle.createFixtureLoader

describe('UniswapV2Factory', () => {

  let wallet: Wallet, other: Wallet

  let factory: UniswapV2Factory
  let pairBytecode: string
  const fixture = async () => {
    const factoryFactory = await ethers.getContractFactory('UniswapV2Factory')
    return (await factoryFactory.deploy(wallet.address)) as UniswapV2Factory
  }

  // What's the meanings of the below code?
  let loadFixture: ReturnType<typeof createFixtureLoader>
  before('create fixture loader', async () => {
    [wallet, other] = await (ethers as any).getSigners()

    loadFixture = createFixtureLoader([wallet, other])
  })

  before('load pair bytecode', async () => {
    // const initCode  = (await ethers.getContractFactory('UniswapV2Pair')).bytecode

    // console.log("initCode");
    // console.log(initCode);
    // // ?? doing contract address not equal?
    // // // Encode the constructor parameters
    // const encodedConstructorParameters = ethers.utils.defaultAbiCoder.encode(['string', 'string'], ['Uniswap V2', 'UNI-V2']);
    // // console.log("pairBytecode2",pairBytecode2);

    // pairBytecode = utils.solidityPack(
    //   ['bytes','bytes'],
    //   [initCode,
    //     encodedConstructorParameters
    //   ]
    // )

    // todo, perhaps the typescript calculate the bytecode have some problem?
    pairBytecode = `${foundry_UniswapV2Pair.bytecode.object}`

  })

  beforeEach('deploy factory', async () => {
    factory = await loadFixture(fixture)
  })


  // it('factory bytecode size', async () => {
  //   // expect(((await waffle.provider.getCode(factory.address)).length - 2) / 2).to.matchSnapshot()
  // })


  it('pair bytecode size', async () => {
    
    await factory.createPair(TEST_ADDRESSES[0], TEST_ADDRESSES[1])
    const pairAddress = getCreate2Address(factory.address, TEST_ADDRESSES,pairBytecode)
    expect(((await waffle.provider.getCode(pairAddress)).length - 2) / 2).to.matchSnapshot()
  })

  it('feeTo, feeToSetter, allPairsLength', async () => {
    expect(await factory.feeTo()).to.eq(constants.AddressZero)
    expect(await factory.feeToSetter()).to.eq(wallet.address)
    expect(await factory.allPairsLength()).to.eq(0)
  })


  // the below can't work correctly, but the forge test OK, now just skip TODO
  // tip: each time the generate address is different, but for my understanding ,it shouldn't be same. perhaps the reson is related with the hardhat environment
  async function createPair(tokens: [string, string]) {
    // console.log("pairBytecode",pairBytecode)
    let create2Address = getCreate2Address(factory.address, tokens, pairBytecode)
    console.log("create2Address",create2Address);

    await expect(factory.createPair(...tokens))
      .to.emit(factory, 'PairCreated')
      // .withArgs(TEST_ADDRESSES[0], TEST_ADDRESSES[1], "0xC79E4a34465a64f12BC64015A2B4163Bb3A6d274", BigNumber.from(1))  


    const address_5 =  await factory.allPairs(0) 
    console.log("address_5",address_5) 

    await expect(factory.createPair(tokens[0], tokens[1])).to.be.reverted  
    await expect(factory.createPair(tokens[1], tokens[0])).to.be.reverted  

    expect(await factory.getPair(tokens[0], tokens[1]), 'getPool in order').to.eq(address_5)
    expect(await factory.getPair(tokens[1], tokens[0]), 'getPool in reverse').to.eq(address_5)

    
    expect(await factory.allPairs(0)).to.eq(address_5)
    expect(await factory.allPairsLength()).to.eq(1)


    const pairContractFactory = await ethers.getContractFactory('UniswapV2Pair')
    const pair = pairContractFactory.attach(address_5)

    expect(await pair.factory()).to.eq(factory.address)
    expect(await pair.token0()).to.eq(TEST_ADDRESSES[0])
    expect(await pair.token1()).to.eq(TEST_ADDRESSES[1])
  }

  it('createPair', async () => {
    await createPair(TEST_ADDRESSES)
  })

  it('createPair:reverse', async () => {
    await createPair(TEST_ADDRESSES.slice().reverse() as [string, string])
  })

  it('createPair:gas', async () => {
    const tx = await factory.createPair(...TEST_ADDRESSES)
    const receipt = await tx.wait()
    // expect(receipt.gasUsed).to.eq(2512920)
  })

  // check the below right or not?
  // it('createPair:gas', async () => {
  //   await snapshotGasCost(factory.createPair(...TEST_ADDRESSES))
  // })

  it('setFeeTo', async () => {
    await expect(factory.connect(other).setFeeTo(other.address)).to.be.revertedWith('UniswapV2: FORBIDDEN')
    await factory.setFeeTo(wallet.address)
    expect(await factory.feeTo()).to.eq(wallet.address)
  })

  it('setFeeToSetter', async () => {
    await expect(factory.connect(other).setFeeToSetter(other.address)).to.be.revertedWith('UniswapV2: FORBIDDEN')
    await factory.setFeeToSetter(other.address)
    expect(await factory.feeToSetter()).to.eq(other.address)
    await expect(factory.setFeeToSetter(wallet.address)).to.be.revertedWith('UniswapV2: FORBIDDEN')
  })
})
