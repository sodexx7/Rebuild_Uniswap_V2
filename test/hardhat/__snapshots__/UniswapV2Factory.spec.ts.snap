// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniswapV2Factory createPair:gas 1`] = `2936691`;

exports[`UniswapV2Factory factory bytecode size 1`] = `15946`;

exports[`UniswapV2Factory pair bytecode size 1`] = `0`;
